### PR TITLE
Fix resolving globally installed npm packages

### PR DESCRIPTION
**Pull request by remcohaszing**
I expected that I would be able to load modules installed with `npm install --global` when passing the `global` option to `loadPlugin`, but it doesn’t work.

My global `.npmrc`:

```ini
audit=false
cache=/home/remco/.cache/npm
init-author-email=remcohaszing@gmail.com
init-author-name=Remco Haszing
init-license=MIT
prefix=/home/remco/.local
update-notifier=false
```

```sh
$ npm ls --global
/home/remco/.local/lib
├── @arethetypeswrong/cli@0.13.10
├── @reporters/testwatch@1.4.3
├── bun@1.0.26
├── httpster@1.2.1
├── react-devtools@5.0.0
├── tsx@4.7.1
├── webpack-bundle-analyzer@4.10.1
└── yarn-deduplicate@6.0.2
```

```
$ node           
Welcome to Node.js v18.19.1.
Type ".help" for more information.
> const { loadPlugin } = await import('load-plugin')
[Module: null prototype] {
  loadPlugin: [AsyncFunction: loadPlugin],
  resolvePlugin: [AsyncFunction: resolvePlugin]
}
> await loadPlugin('tsx', {global: true})
Uncaught:
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'tsx' imported from /usr/lib/node_modules
    at __node_internal_ (file:///home/remco/Projects/load-plugin/node_modules/import-meta-resolve/lib/errors.js:431:11)
    at new NodeError (file:///home/remco/Projects/load-plugin/node_modules/import-meta-resolve/lib/errors.js:374:5)
    at packageResolve (file:///home/remco/Projects/load-plugin/node_modules/import-meta-resolve/lib/resolve.js:1038:9)
    at moduleResolve (file:///home/remco/Projects/load-plugin/node_modules/import-meta-resolve/lib/resolve.js:1101:20)
    at defaultResolve (file:///home/remco/Projects/load-plugin/node_modules/import-meta-resolve/lib/resolve.js:1266:15)
    at resolve (file:///home/remco/Projects/load-plugin/node_modules/import-meta-resolve/index.js:32:12)
    at resolvePlugin (file:///home/remco/Projects/load-plugin/lib/index.js:205:14)
    at Module.loadPlugin (file:///home/remco/Projects/load-plugin/lib/index.js:123:22)
    at REPL2:1:48 {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

I managed to make this work by replacing [`@npmcli/config`](https://github.com/npm/cli/tree/latest/workspaces/config) with [`global-directory`](https://github.com/sindresorhus/global-directory). This also removes the dependency on `import.meta.url`, which is still a reason this package needs hacks to be usable when compiled to CJS (#16). As a bonus it has much fewer indirect dependencies.

The broken test relies on a detail on the environment it runs in. If these changes are wanted, I intend to actually install a package globally in CI instead to provide a more robust test case.

### CODE REVIEW

**Comment by wooorm** [comment]
> This also removes the dependency on import.meta.url, which is still a reason this package needs hacks to be usable when compiled to CJS (https://github.com/wooorm/load-plugin/pull/16)

How is #16 related to this, that was about an `await` on `fs.stat`? What hacks are you applying? `import-meta-resolve` / `import.meta.resolve` are sync now?

> My global .npmrc:

How does npm know you want things in `.local`?
I think you need to set a `prefix` variable or so in your env file or some other variable?
What does `config.loadGlobalPrefix` give on your system?
Is your folder found somewhere else on `config`?

> As a bonus it has much fewer indirect dependencies.

But it’s always going to do things differently than how npm actually works.
When npm makes a change, it’ll be a while for it to be updated.
Feels weird to depend on npm but use something that models how npm works if we can also use npm?

**Comment by remcohaszing** [comment]
> > This also removes the dependency on import.meta.url, which is still a reason this package needs hacks to be usable when compiled to CJS (#16)
> 
> How is #16 related to this, that was about an `await` on `fs.stat`? What hacks are you applying? `import-meta-resolve` / `import.meta.resolve` are sync now?

`load-plugin` uses `import.meta.url`. This is not available in CJS. So getting rid of the top-level await fixed one issue, but then another arose.

> > My global .npmrc:
> 
> How does npm know you want things in `.local`?

Because of the `prefix` variable in the `.npmrc` file I posted.

> I think you need to set a `prefix` variable or so in your env file or some other variable?

I think you mean the `prefix` variable in the `.npmrc` I posted?

> What does `config.loadGlobalPrefix` give on your system? Is your folder found somewhere else on `config`?

Good question. It gives me this:

```
Config {
  definitions: {},
  deprecated: {},
  types: {},
  shorthands: undefined,
  defaults: {},
  npmPath: '/home/remco/Projects/load-plugin/lib/',
  npmBin: '/home/remco/Projects/load-plugin/lib/bin/npm-cli.js',
  argv: [ '/usr/bin/node', '/home/remco/Projects/load-plugin/lib/index.js' ],
  env: { /* truncated */ },
  execPath: '/usr/bin/node',
  platform: 'linux',
  cwd: '/home/remco/Projects/load-plugin',
  excludeNpmCwd: false,
  globalPrefix: '/usr',
  localPrefix: null,
  localPackage: null,
  home: null,
  data: Map(7) {
    'default' => ConfigData { [Symbol(valid)]: true },
    'builtin' => ConfigData { [Symbol(valid)]: true },
    'global' => ConfigData { [Symbol(valid)]: true },
    'user' => ConfigData { [Symbol(valid)]: true },
    'project' => ConfigData { [Symbol(valid)]: true },
    'env' => ConfigData { [Symbol(valid)]: true },
    'cli' => ConfigData { [Symbol(valid)]: true },
    set: [Function (anonymous)],
    delete: [Function (anonymous)]
  },
  sources: Map(0) {},
  list: [
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <[Object: null prototype] {}> {},
    [Object: null prototype] {}
  ]
}
```

So it seems to be completely unaware of the npm configuration, and it points to `npmPath` and `npmBin` that don’t exist. `globalPrefix` points to the prefix where Node.js was installed. This is the default npm prefix too.

**Comment by wooorm** [comment]
~~It uses `import-meta-resolve`. Which is a package that I made~~. Right it also uses `import.meta.url`, which, how does that now work when bundled? CJS is not a goal of this/my projects?

> Because of the prefix variable in the .npmrc file I posted.

Ah I missed that you had one. Yes, I think this should just work. 

> So it seems to be completely unaware of the npm configuration, and it points to npmPath and npmBin that don’t exist. globalPrefix points to the prefix where Node.js was installed. This is the default npm prefix too.

Is your `.npmrc` itself in a weird place? 🤔 It seems `npm ls --global` finds it but `@npmcli/config` as it’s used right now doesn’t. Perhaps `@npmcli/config` needs to be configured?

https://github.com/wooorm/load-plugin/pull/18/files#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1L50-L53

What is the value currently if you log the `npmPath` passed there? Could it be that your `.npmrc` is not in one of its parent folders? 🤔

**Comment by wooorm** [comment]
Ohhhh. Wait. The `npmPath` points to where `load-plugin` currently is installed; and that’s somewhere else for you? How did you install `load-plugin`?

**Comment by wooorm** [comment]
Here’s what I get:

```txt
Config {
  definitions: {},
  deprecated: {},
  types: {},
  shorthands: undefined,
  defaults: {},
  npmPath: '/Users/tilde/Projects/oss/load-plugin/lib/',
  npmBin: '/Users/tilde/Projects/oss/load-plugin/lib/bin/npm-cli.js',
  argv: [ '/Users/tilde/.nvm/versions/node/v21.2.0/bin/node', '/Users/tilde/Projects/oss/load-plugin/test' ],
  env: { /* truncated */ },
  execPath: '/Users/tilde/.nvm/versions/node/v21.2.0/bin/node',
  platform: 'darwin',
  cwd: '/Users/tilde/Projects/oss/load-plugin',
  excludeNpmCwd: false,
  globalPrefix: '/Users/tilde/.nvm/versions/node/v21.2.0',
  localPrefix: null,
  localPackage: null,
  home: null,
  data: Map(7) {
    'default' => ConfigData { [Symbol(valid)]: true },
    'builtin' => ConfigData { [Symbol(valid)]: true },
    'global' => ConfigData { [Symbol(valid)]: true },
    'user' => ConfigData { [Symbol(valid)]: true },
    'project' => ConfigData { [Symbol(valid)]: true },
    'env' => ConfigData { [Symbol(valid)]: true },
    'cli' => ConfigData { [Symbol(valid)]: true },
    set: [Function (anonymous)],
    delete: [Function (anonymous)]
  },
  sources: Map(0) {},
  list: [
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <Object <Complex prototype>> {},
    Object <[Object: null prototype] {}> {},
    [Object: null prototype] {}
  ]
}
```

**Comment by wooorm** [comment]
What happens if you pass `npmPath: process.cwd()` or, could you experiment with if changing that path a bit works for you?

**Comment by remcohaszing** [comment]
`npmPath` is the npm installation `@npmcli/config` is part of. It’s not really used to resolve `.npmrc`, except the `.npmrc` file in the npm installation folder is loaded as well.

This shows how you’re supposed to use it https://github.com/npm/cli/blob/v10.4.0/workspaces/config/test/index.js#L1101-L1116

```js
const argv = [
  'node',
  __filename,
  `--prefix=${path}`,
  `--userconfig=${path}/.npmrc`,
  `--globalconfig=${path}/etc/npmrc`,
]
const opts = {
  shorthands: {},
  argv,
  env: {},
  definitions: {
    registry: { default: registry },
  },
  npmPath: process.cwd(),
}
```

So basically it looks like you’re supposed to resolve `.npmrc` yourself. Though I am a bit puzzled by all the indirection used in the `Config` class.

In the end both `@npmcli/config` and `global-directory` load the contents with the `ini` package.

**Comment by wooorm** [comment]
How do you want to proceed?

It’s been a while so I forgot some details.
But I’d likely still prefer to use `@npmcli/config` instead of `global-directory`, so that we can call npm code directly and then if they ever change, it would be auto-included here.

**Comment by remcohaszing** [comment]
I fiddled a bit again after giving this some rest, and I managed to make it work.

- `npmPath` is completely redundant for us. We can set it to an empty string. This solves the CJS compatibility issue.
- We need to pass actual `definitions` imported from `@npmcli/config/lib/definitions/definitions.js`, not some stub.
- We need to use `conf.load()` instead of `conf.loadGlobalPrefix()`. This returns a promise though, so awaiting it top-level again poses a CJS compatibility issue.
- To solve this compatibility issue, we can move all asynchronous preparation logic into `resolvePlugin()`.

Also we should pass `argv: []` to the `Config` constructor to prevent it from using `process.argv`. I don’t think we want that in a library.

In short, this works:

```js
import Config from '@npmcli/config'
import definitions from '@npmcli/config/lib/definitions/definitions.js'

const conf = new Config({
  argv: [],
  definitions,
  npmPath: ''
})

await conf.load()

console.log(conf.globalPrefix)
```

I can go ahead and update this PR to apply those changes.

Still though, These are the pros of `global-directory`:

- Much fewer dependencies
- Simpler to use
- Well documented
- Type definitions

VS the pros of `@npmcli/config`:

- It’s official

Additionally, `global-directory` supports Yarn and has an open issue to support pnpm. I see value in supporting these additionally.

**Comment by wooorm** [comment]
Released in [`6.0.3`](https://github.com/wooorm/load-plugin/releases/tag/6.0.3)!
